### ISSUE TITLE
[Question] Reducing repeated information with Unions

### ISSUE THREAD

**Issue by Jazcash** [sinclairzx81/typebox#113]
I want to write a schema that describes an object that contains a password and _either_ a username or an email (or both).

```
const test = Type.Union([
    Type.Object({
        password: Type.String(),
        email: Type.String(),
    }),
    Type.Object({
        password: Type.String(),
        username: Type.String(),
    }),
]);
```

This achieves what I want, but I wondered if there was a way to write this that doesn't require writing the definition for `password` twice?

**Comment by Jazcash** [comment]
```
const test = Type.Intersect([
    Type.Object({
        password: Type.String(),
    }),
    Type.Union([
        Type.Object({
            username: Type.String()
        }),
        Type.Object({
            email: Type.String()
        }),
    ])
]);
```

I thought this would work but seemingly not

**Comment by sinclairzx81** [comment]
@Jazcash Hi. There are a few ways to approach this. See example below that uses both Unions and Intersections to achieve this.

Example also available [here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4GUYCGMwAxnAL5wBmUEIcA5AAIDOwAdiQDYHBQD0MdBgBGEAB4MAsAChZ-fnAC0K1WvUbNW7Tt3r5igPJhiEdnACCALjgBRcTAzsAJnBJ0QZuGFqZYqRjACFhYAd2hnBjhQ4BgACzgYcLgCZ2dY4DMCLkThFgNlPSLiktLZBFk4NzMWeABhKAxnJ2JsqvaOgF4UYQA6QxEAKwwSGAAKJCCQ8KhnGzRMXvwoDgBzMYBKCg3ZSur2WrgAVRYMKAA5AhAMBqaW4DbuhYxegEl2RyhT0bGAbVvmh8HlwcM9+kMRuMkABXU5QdhXDDzPrLNabbYAXR2chkVXcB3gthAvC4APubXaTz670+33G-0agNaIJ6iwGwx+SAwxOAXGRi1R7HWW3IGyxu1x1GhnFM5lhZ0u1zJQOyyAwtTGJEZ5L5eEIxBIAB4hJgIFRjnDFTdtSquAA+LYVSV4mrwGFwhHXHBTMIRChwbpau62uAKRLqmB7ch7KjS0aZczcknK5lqjVBpnAmz4IikY3CM12Hmkm3Mh2IPYugmIOBJ3ne4K+2b+wOl4GhxSOWpR2TRmRh0qDofDpQFYyyuAAIRsR3YCdZ6uisQShFWqyacGaWtAHCI0AKI8PR605T2+MOJwViJT7apbIhnMruUwNg6YIAMrEztkxgx5fDESUDMdQYDYsCff9PSRBclhgFYhU2cDnW8RsZjmGDBWFKNsTPV0i2TNtHhg9lIQmJ8TWgt8+k-T4fwYOsuCAwiuFApCOgY18qjBTDEKfH00P5F4eOxKpRVw6sbyIsFZwTP5Lwua9mJwIkCODZlxRxKpYxledgNtNNxj05ls31PMKMLST7UdJ8wliEg4k1Zjego6zkLxYIMEYSDAKM4EGBsJ0Og6c83TgbyvRQ6Y-UoVs1PbMMu0jNz2hERoCAAayfUSsrcDzGAYpi4uyfyK2SqtDi5YsGyi5sYrcZiO3DbsyrgVKMAynK+w6LqKF7IA)

```typescript
import { Type, Static } from '@sinclair/typebox'

// -----------------------------------------------------------------------
// Option A: Extend common property 'password' with two additional types
// -----------------------------------------------------------------------
{
  const Credential         = Type.Object({ password: Type.String() })

  const UserNameCredential = Type.Intersect([Credential, Type.Object({ username: Type.String() })])

  const EmailCredential    = Type.Intersect([Credential, Type.Object({ email: Type.String() })])

  function userNameCredentialTest(credential: Static<typeof UserNameCredential>) {
    const { username, password } = credential // test
  }
  function emailCredentialTest(credential: Static<typeof EmailCredential>) {
    const { email, password } = credential // test
  }
}
// -----------------------------------------------------------------------
// Option B: Union Types with tagged decriminator
// -----------------------------------------------------------------------
{
  const UserNameCredential = Type.Object({
    type:     Type.Literal('username-credential'),
    username: Type.String(),
    password: Type.String()
  })

  const EmailCredential = Type.Object({
    type:     Type.Literal('email-credential'),
    email:    Type.String(),
    password: Type.String()
  })
  const Credential = Type.Union([UserNameCredential, EmailCredential])

  function credentialTest(credential: Static<typeof Credential>) {
    switch(credential.type) {
      case 'username-credential': {
        const { username, password } = credential // test
        break
      }
      case 'email-credential': {
        const { email, password } = credential // test
        break
      }
    }
  }
}
```
Hope that helps
S

**Comment by Jazcash** [comment]
Thanks for taking the time, but unfortunately option A is two separate types and I'm only looking for one. Option B is what I currently have and what I wrote in my first code example, but requires writing the definition for `password` twice. It's not a massive deal but I may need to reuse this type of structure a lot so something like my second example would be ideal, but yeah, just not sure why the Static doesn't evaluate to the expected type there but the outputted JSON schema does

**Comment by sinclairzx81** [comment]
@Jazcash Hi, this is likely related to issue https://github.com/sinclairzx81/typebox/issues/102#issuecomment-914404211, where there are currently difficulties capturing operator precedence (specifically enclosed `&` and `|`) in these kind of inference scenarios.

I am looking into solutions to this, so may be able to provide an update at some point down the road. Will leave this issue open for the time being.

Cheers
S

**Comment by Jazcash** [comment]
Been trying to work out how `Static` works and man is that stuff scary. Probably of no use to you but just been playing around and found a cool way to do intersections with the variadic tuple type stuff in 4.0

```
type Unionize<T extends readonly unknown[]> = T[number];
type Intersect<T extends readonly unknown[]> = T extends readonly [infer U, ...infer Rest] ? (U & Intersect<Rest>) : {};

const Password = { password: "" };
const Email = { email: "" };
const Username = { username: "" };

type Out = Intersect<[typeof Password, Unionize<[typeof Email, typeof Username]>]>;

// @ts-expect-error
const a: Out = { };
// @ts-expect-error
const b: Out = { stuff: "bob" };
const c: Out = { password: "", email: "" };
const d: Out = { password: "", username: "" };
const e: Out = { password: "", username: "", email: "" };
```
[Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAqgdgSwPaIF4QDwBUoQB7ARwAmAzlAE4QCGxKANiFAK5wDWcSA7nANoC6APigBeKFl5xmAWwBGECvwDcAWABQoSFACScQhVIQAxsGy4CRMpRp04jFu048BwsTnyES5KrQZNeCHAAZgqwADRQAHTRgSEUUABKEKTA-FAA-FAAFDBQAGQ6egqGJhhJKYIAlFAAXFAA3gC+qmrqRigpUAAK1KSkXEgUxKINUGC9-YPEdQBEM1DNbR3AUACi0tQI9CP1uBtbs-OLau1wnTCGFHDU0tBiu8yX17eHCy3qmtAA8swrYrr6EqmXifJBBboTAZDCLwZBoTAg8AQMFrfb0CKg8EXBTPCBCITvNQAeiJUAAAsBSABafCQEw0igUQZLM4rah1H5-UbHEnkyk0vB04AMpkUFmdWQc347KApZhBIKzWRIWRHFqnTpGKVc3bjPpQ6ZQOYRCBo17HDUrQ2cmV6yZDQ4RR44m4Qc3q5a4bW2yFTR0sJ6u-2mzb0d1AA)

**Comment by sinclairzx81** [comment]
@Jazcash Hi, Have just published an update on `@sinclair/typebox@0.21.2` that resolves this issue. I've made mention in the changelog about the update which you can read about below. 

This work was undertaken to align with various changes in TS 4.5.

ChangeLog [link](https://github.com/sinclairzx81/typebox/blob/master/changelog.md)

TypeScript Playground [link](https://www.typescriptlang.org/play?ts=4.5.0-beta#code/JYWwDg9gTgLgBAbzgZRgQxsAxgGjgFQE8wBTOAXzgDMoIQ4AiAAQGdgA7LAGzWCgHoYxEgCMIADwYBuALAAoeVgjsW8AIJwAvAWEA6APIiAViSwwAFEjQAuHaV2ooHAObmAlBTeLlquACEtOxIDY1MLJBFbIntHF3dPbxV4AGFA6ODDEzNLOCwovVj2Vw9yLwU5JSSCNL0ASXYYEigWMPMAbTU8dN0AVXZgZXa-PGSAXTdx+SFSau1UDGwAHmmSCCoCAD55eX5+OBXZxDgbOFUnIoo4ADI4HMjTmHPnS4AfI7yHp4TyoA)

Will be happy to close off this issue at last!
Cheers
S

**Comment by Jazcash** [comment]
Awesome stuff, works great!
